packages/data-stores/src/auth/actions.ts

```diff
@@ -126,48 +127,94 @@ export function createActions( {
 		const username = yield select( STORE_KEY, 'getUsernameOrEmail' );
 
 		try {
-			const loginResponse = yield fetchAndParse(
-				// TODO Wrap this in `localizeUrl` from lib/i18n-utils
-				'https://wordpress.com/wp-login.php?action=login-endpoint',
-				{
-					credentials: 'include',
-					method: 'POST',
-					headers: {
-						Accept: 'application/json',
-						'Content-Type': 'application/x-www-form-urlencoded',
-					},
-					body: stringify( {
-						remember_me: true,
-						username,
-						password,
-						client_id,
-						client_secret,
-					} ),
-				}
-			);
+			let loginResponse = yield* wpLogin( 'login-endpoint', {
+				remember_me: true,
+				username,
+				password,
+			} );
+
+			if ( ! loginResponse.success ) {
+				return receiveWpLoginFailed( loginResponse );
+			}
+
+			if ( loginResponse.data.two_step_notification_sent ) {
+				yield receiveWpLogin( loginResponse );
 
-			if ( loginResponse.ok && loginResponse.body.success ) {
-				if ( loadCookiesAfterLogin ) {
-					yield reloadProxy();
+				if ( loginResponse.data.two_step_notification_sent === 'push' ) {
+					loginResponse = yield* handlePush2fa( loginResponse );
 				}
-				yield remoteLoginUser( loginResponse.body.data.token_links );
-				yield receiveWpLogin( loginResponse.body );
-			} else {
-				yield receiveWpLoginFailed( loginResponse.body );
 			}
+
+			yield* handleSuccessfulLogin( loginResponse );
+
+			return;
 		} catch ( e ) {
 			const error = {
 				code: e.name,
 				message: e.message,
 			};
 
-			yield receiveWpLoginFailed( {
+			return receiveWpLoginFailed( {
 				success: false,
 				data: { errors: [ error ] },
 			} );
 		}
 	}
 
+	function* handleSuccessfulLogin( response: WpLoginSuccessResponse ) {
+		if ( loadCookiesAfterLogin ) {
+			yield reloadProxy();
+		}
+		yield remoteLoginUser( response.data.token_links || [] );
+		yield receiveWpLogin( response );
+	}
+
+	function* handlePush2fa( response: WpLoginSuccessResponse ) {
+		const { user_id, push_web_token } = response.data;
+		let { two_step_nonce_push: two_step_nonce } = response.data;
+
+		while ( ! response.data.token_links ) {
+			response = yield wpLogin( 'two-step-authentication-endpoint', {
+				remember_me: true,
+				auth_type: 'push',
+				user_id,
+				two_step_nonce,
+				two_step_push_token: push_web_token,
+			} );
+
+			if ( ! response.success ) {
```

Comment:
There should also be error cases which indicate that the push authentication has really failed an will never succeed. I think you should be more specific with this check and then handle any other error's seperately

Reply:
Agree, I reckon this goes hand in hand with timing out too.